### PR TITLE
HWKAPM-491 Fix startup issue where APM server is not currently availa…

### PR DESCRIPTION
**Pull request by objectiser**
…ble, to ensure that the client environment retries to get the configuration.

### CODE REVIEW

**Comment by objectiser** [comment]
Review notes:
* This change has primarily been made to fix an issue where the APM server is not available when a client environment is started up, which causes it to fail obtaining the configuration and therefore not report any trace fragments, even after the APM server has been started up. So the change is basically to add a retry if it fails to obtain the collector config.
* This issue was found when using APM within the Openshift CDK - when initially deploying the APM server and an instrumented client it worked fine, but when the VMs were restarted (as it started the APM server and instrumented client at the same time) the APM server was not fully started at the point where the instrumented client was trying to get the collector config. In production this situation should not occur, but should be handled in any case.

**Comment by objectiser** [comment]
@pavolloffay If you have time, could you review this PR? If not let me know and I will ask someone else, as I know you are on a well deserved break :)

File: client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java [hunk 1]
```diff
@@ -39,6 +43,9 @@
  */
 public class ClientManager {
 
+    /**  */
+    private static final int DEFAULT_CONFIG_RETRY_INTERVAL = 10;
```
Comment (pilhuhn) [inline client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java]: What is the unit?

File: client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java [hunk 2]
```diff
@@ -76,24 +86,40 @@ public static void initialize(Retransformer trans) {
...
 
+    protected static void initConfig() {
+        // Read configuration
+        CollectorConfiguration config = configService.getCollector(null, null, null, null);
+
+        if (config != null) {
+            try {
+                updateInstrumentation(config);
+            } catch (Exception e) {
+                log.severe("Failed to update instrumentation rules: " + e);
```
Comment (pilhuhn) [inline client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java]: Can you continue here? Don't you want to bubble up the exception or even directly bail out?
If bubbling up, then don't log+throw

File: client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java (hunk 1 above)
Comment (pavolloffay) [inline client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java]: I have the same question, apparently It is in seconds.

**Comment by pavolloffay** [comment]
Btw I'm here until Thursday, so I can review

File: client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java (hunk 2 above)
Comment (objectiser) [inline client/instrumenter/src/main/java/org/hawkular/apm/instrumenter/ClientManager.java]: This code executes in the javaagent - so throwing an exception would prevent the application from starting up. Not sure if that is desirable?
Currently only log severe message - but have a separate jira regarding logging admin issues to the server, so possibly this error could be sent to the server at that point, where an operator could take action.

File: client/collector/src/main/java/org/hawkular/apm/client/collector/DefaultTraceCollector.java [hunk 3]
```diff
@@ -117,56 +131,77 @@ public void setConfigurationService(ConfigurationService cs) {
...
+                                    log.finer("Removing config for btxn '" + stringBusinessTxnConfigEntry.getKey() + "' = " + btc);
                                 }
 
-                                if (btc.getLastUpdated() > configLastUpdated) {
-                                    configLastUpdated = btc.getLastUpdated();
+                                filterManager.remove(stringBusinessTxnConfigEntry.getKey());
+                                processorManager.remove(stringBusinessTxnConfigEntry.getKey());
+                            } else {
+                                if (log.isLoggable(Level.FINER)) {
+                                    log.finer("Changed config for btxn '" + stringBusinessTxnConfigEntry.getKey() + "' = " + btc);
```
Comment (pavolloffay) [inline client/collector/src/main/java/org/hawkular/apm/client/collector/DefaultTraceCollector.java]: This log could be there only once, `btc.toString()` contains deleted attribute. On the other hand explicit logging might be better for redibility of the log file.

Comment (objectiser) [inline client/collector/src/main/java/org/hawkular/apm/client/collector/DefaultTraceCollector.java]: Sorry Pavol not sure I understand? What do you mean about btc.toString() containing deleted attributes?

**Comment by objectiser** [comment]
Thanks @pilhuhn and @pavolloffay

**Pull request merged**